espnet2/enh/layers/beamformer_th.py

```diff
@@ -0,0 +1,1065 @@
+            psd_speech,
+            psd_noise,
+            reference_vector,
+            n_iter=iterations,
+            diagonal_loading=False,
+        )
+    elif mode == "evd":
+        rtf = torchaudio.functional.rtf_evd(psd_speech)
+    else:
+        raise ValueError("Unknown mode: %s" % mode)
+    return rtf
+
+
+def get_mvdr_vector(
+    psd_s,
+    psd_n,
+    reference_vector: Union[torch.Tensor, int],
+    diagonal_loading: bool = True,
+    diag_eps: float = 1e-7,
+    eps: float = 1e-8,
+):
+    """Return the MVDR (Minimum Variance Distortionless Response) vector:
+
+        h = (Npsd^-1 @ Spsd) / (Tr(Npsd^-1 @ Spsd)) @ u
+
+    Reference:
+        On optimal frequency-domain multichannel linear filtering
+        for noise reduction; M. Souden et al., 2010;
+        https://ieeexplore.ieee.org/document/5089420
+
+    Args:
+        psd_s (torch.complex64):
+            speech covariance matrix (..., F, C, C)
+        psd_n (torch.complex64):
+            observation/noise covariance matrix (..., F, C, C)
+        reference_vector (torch.Tensor): (..., C) or an integer
+        diagonal_loading (bool): Whether to add a tiny term to the diagonal of psd_n
+        diag_eps (float):
+        eps (float):
+    Returns:
+        beamform_vector (torch.complex64): (..., F, C)
+    """  # noqa: D400
+    return torchaudio.functional.mvdr_weights_souden(
+        psd_s,
+        psd_n,
+        reference_vector,
+        diagonal_loading=diagonal_loading,
+        diag_eps=diag_eps,
+        eps=eps,
+    )
+
+
+def get_mvdr_vector_with_rtf(
+    psd_n: torch.Tensor,
+    psd_speech: torch.Tensor,
+    psd_noise: torch.Tensor,
+    iterations: int = 3,
+    reference_vector: Union[int, torch.Tensor, None] = None,
+    diagonal_loading: bool = True,
+    diag_eps: float = 1e-7,
+    eps: float = 1e-8,
+) -> torch.Tensor:
+    """Return the MVDR (Minimum Variance Distortionless Response) vector
+        calculated with RTF:
+
+        h = (Npsd^-1 @ rtf) / (rtf^H @ Npsd^-1 @ rtf)
+
+    Reference:
+        On optimal frequency-domain multichannel linear filtering
+        for noise reduction; M. Souden et al., 2010;
+        https://ieeexplore.ieee.org/document/5089420
+
+    Args:
+        psd_n (torch.complex64):
+            observation/noise covariance matrix (..., F, C, C)
+        psd_speech (torch.complex64):
+            speech covariance matrix (..., F, C, C)
+        psd_noise (torch.complex64):
+            noise covariance matrix (..., F, C, C)
+        iterations (int): number of iterations in power method
+        reference_vector (torch.Tensor or int): (..., C) or scalar
+        diagonal_loading (bool): Whether to add a tiny term to the diagonal of psd_n
+        diag_eps (float):
+        eps (float):
+    Returns:
+        beamform_vector (torch.complex64): (..., F, C)
+    """  # noqa: H405, D205, D400
+    if diagonal_loading:
+        psd_noise = tik_reg(psd_noise, reg=diag_eps, eps=eps)
```

Comment:
diagonal loading is included in [`torchaudio.functional.mvdr_weights_rtf`](https://pytorch.org/audio/stable/_modules/torchaudio/functional/functional.html#mvdr_weights_rtf), so it can be removed.

Reply:
Thanks! I now remove this line and use torchaudio's implementation. For the use of `tik_reg` in some other functions, I intended to do explicit diagonal loading to avoid repeating this operation in multiple function calls.